### PR TITLE
feat: display warning if patch_sklearn() called after sklearn imports

### PR DESCRIPTION
**Pull request by ahuber21**

### CODE REVIEW

File: sklearnex/dispatcher.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 import os
 import sys
+import warnings
```
Comment (napetrov) [inline sklearnex/dispatcher.py:19]: i guess we need to use unified mechanism with logging ?

File: sklearnex/dispatcher.py [hunk 2]
```diff
@@ -289,6 +292,14 @@ def patch_sklearn(name=None, verbose=True, global_patch=False, preview=False):
             "for scikit-learn >= 0.22 only ..."
         )
 
+    if not no_msg and "sklearn" in sys.modules.keys():
```
Comment (napetrov) [inline sklearnex/dispatcher.py:295]: i would say would worth printing out submodules list that we see already imported from scikit?

File: sklearnex/dispatcher.py (hunk 1 above)
Comment (ahuber21) [reply sklearnex/dispatcher.py:19]: Would you prefer logging a warning to the `sklearnex` logger? We're not using logging anywhere so far in this part.

File: sklearnex/dispatcher.py (hunk 2 above)
Comment (ahuber21) [reply sklearnex/dispatcher.py:295]: I did it locally and it's always a pretty long list, not so sure if that's useful.

```
$ python -c "import json,sys;from sklearn.linear_model import LogisticRegression;print(json.dumps(list(sys.modules.keys()), indent=2, sort_keys=True))" | grep sklearn                                         
  "sklearn._config",                                             
  "sklearn.__check_build._check_build",
  "sklearn.__check_build",                              
  "sklearn._distributor_init",                                          
  "sklearn.exceptions", 
  "sklearn.utils._joblib", 
  "sklearn.utils._bunch",                                  
  "sklearn.utils._metadata_requests",                                
  "sklearn.utils.metadata_routing",
  "sklearn.utils._estimator_html_repr",                             
  "sklearn.externals",                                        
  "sklearn.externals._packaging",                 
  "sklearn.externals._packaging._structures",
  "sklearn.externals._packaging.version",
  "sklearn.utils.deprecation",            
  "sklearn.utils.fixes",    
  "sklearn.utils._array_api",       
  "sklearn.utils._isfinite",
  "sklearn.utils.validation",              
  "sklearn.utils._param_validation",
  "sklearn.utils.class_weight",
  "sklearn.utils.discovery",        
  "sklearn.utils.murmurhash",                    
  "sklearn.utils",                   
  "sklearn.utils._available_if",    
  "sklearn.utils._set_output",  
  "sklearn.utils._tags",    
  "sklearn.base",   
  "sklearn.utils._openmp_helpers", 
  "sklearn.utils._show_versions",       
  "sklearn",                      
  "sklearn.utils._logistic_sigmoid",
  "sklearn.utils.sparsefuncs_fast", 
  "sklearn.utils.extmath",  
  "sklearn.utils.sparsefuncs",    
  "sklearn.utils._encode",                   
  "sklearn.utils._mask",
  "sklearn.preprocessing._encoders",
  "sklearn.preprocessing._data",
  "sklearn.utils.stats",
  "sklearn.preprocessing._discretization",
  "sklearn.utils.metaestimators",     
  "sklearn.preprocessing._function_transformer",
  "sklearn.utils.multiclass",     
  "sklearn.preprocessing._label",
  "sklearn.preprocessing._csr_polynomial_expansion",
  "sklearn.preprocessing._polynomial",
  "sklearn.preprocessing._target_encoder_fast",
  "sklearn.preprocessing._target_encoder",
  "sklearn.preprocessing", 
  "sklearn.utils._random",
  "sklearn.utils._seq_dataset",
  "sklearn.utils.parallel",
  "sklearn.linear_model._base",
  "sklearn.linear_model._bayes",
  "sklearn.utils._response",     
  "sklearn.utils._plotting",       
  "sklearn.metrics.cluster._bicluster",
  "sklearn.metrics.cluster._expected_mutual_info_fast",
  "sklearn.metrics.cluster._supervised",
  "sklearn.metrics._dist_metrics",
  "sklearn.metrics._pairwise_distances_reduction._datasets_pair",
  "sklearn.utils._cython_blas",              
  "sklearn.metrics._pairwise_distances_reduction._base",
  "sklearn.metrics._pairwise_distances_reduction._middle_term_computer",
  "sklearn.utils._heap",         
  "sklearn.utils._sorting",     
  "sklearn.metrics._pairwise_distances_reduction._argkmin",
  "sklearn.metrics._pairwise_distances_reduction._argkmin_classmode",
  "sklearn.utils._vector_sentinel",
  "sklearn.metrics._pairwise_distances_reduction._radius_neighbors",                                                                                                                                          
  "sklearn.metrics._pairwise_distances_reduction._dispatcher",
"sklearn.metrics._pairwise_distances_reduction",
  "sklearn.metrics._pairwise_fast",
  "sklearn.metrics.pairwise",
  "sklearn.metrics.cluster._unsupervised",
  "sklearn.metrics.cluster",
  "sklearn.metrics._classification",
  "sklearn.metrics._plot",
  "sklearn.metrics._plot.confusion_matrix",
  "sklearn.metrics._base",
  "sklearn.metrics._ranking",
  "sklearn.metrics._plot.det_curve",
  "sklearn.metrics._plot.precision_recall_curve",
  "sklearn.metrics._plot.regression",
  "sklearn.metrics._plot.roc_curve",
  "sklearn.metrics._regression",
  "sklearn.metrics._scorer",
  "sklearn.metrics",
  "sklearn.model_selection._split",
  "sklearn.model_selection._validation",
  "sklearn.model_selection._plot",
  "sklearn.utils.random",
  "sklearn.model_selection._search",
  "sklearn.model_selection",
  "sklearn.linear_model._cd_fast",
  "sklearn.linear_model._coordinate_descent",
  "sklearn._loss._loss",
  "sklearn._loss.link",
  "sklearn._loss.loss",
  "sklearn._loss",
  "sklearn.utils.optimize",
  "sklearn.linear_model._linear_loss",
  "sklearn.linear_model._glm._newton_solver",
  "sklearn.linear_model._glm.glm",
  "sklearn.linear_model._glm",
  "sklearn.linear_model._huber",
  "sklearn.utils.arrayfuncs",
  "sklearn.linear_model._least_angle",
  "sklearn.svm._bounds",
  "sklearn.svm._liblinear",
  "sklearn.svm._libsvm",
  "sklearn.svm._libsvm_sparse",
  "sklearn.svm._base",
  "sklearn.svm._classes",
  "sklearn.svm",
  "sklearn.utils._weight_vector",
  "sklearn.linear_model._sgd_fast",
  "sklearn.linear_model._sag_fast",
  "sklearn.linear_model._sag",
  "sklearn.linear_model._logistic",
  "sklearn.linear_model._omp",
  "sklearn.linear_model._stochastic_gradient",
  "sklearn.linear_model._passive_aggressive",
  "sklearn.linear_model._perceptron",
  "sklearn.linear_model._quantile",
  "sklearn.linear_model._ransac",
  "sklearn.linear_model._ridge",
  "sklearn.linear_model._theil_sen",
  "sklearn.linear_model"
```

File: sklearnex/dispatcher.py (hunk 1 above)
Comment (samir-nasibli) [reply sklearnex/dispatcher.py:19]: I think it make sense to extend sklearnex logger for that. https://github.com/intel/scikit-learn-intelex/blob/86b392d5a62e214765dec324cc8f1453f4dba091/sklearnex/_utils.py#L49

File: sklearnex/dispatcher.py (hunk 2 above)
Comment (icfaust) [inline sklearnex/dispatcher.py:295]: If name is specified, then the user is patching a specific estimator or function. Either the message should apply to only the function/estimator being patched, or the message should not be displayed.

Comment (ahuber21) [reply sklearnex/dispatcher.py:295]: Very good point. I didn't think at all about "non-default" scenarios so far. Are there other "exotic" use cases that you can think of?

**Comment by ahuber21** [comment]
Had to get a little creative, but it seems to be doing fine now. Please let me know what you think @icfaust @napetrov

File: sklearnex/dispatcher.py [hunk 3]
```diff
@@ -280,7 +285,58 @@ def get_patch_names():
     return list(get_patch_map().keys())
 
 
-def patch_sklearn(name=None, verbose=True, global_patch=False, preview=False):
+def check_entity_loaded(
+    name: Optional[str] = None, modules: Optional[dict[str, Any]] = None
```
Comment (ethanglaser) [inline sklearnex/dispatcher.py:289]:
```suggestion
    name: Optional[str] = None, modules: Optional[Dict[str, Any]] = None
```

File: sklearnex/dispatcher.py [hunk 4]
```diff
@@ -14,12 +14,17 @@
 # limitations under the License.
 # ==============================================================================
 
+import logging
 import os
 import sys
+import types
 from functools import lru_cache
+from typing import Any, Optional
```
Comment (ethanglaser) [inline sklearnex/dispatcher.py:22]:
```suggestion
from typing import Any, Dict, Optional
```

File: sklearnex/dispatcher.py [hunk 5]
```diff
@@ -313,7 +318,58 @@ def get_patch_names():
...
+                        cls.__name__
+                        for cls in vars(module).values()
+                        if isinstance(cls, type)
+                    ]
+                )
+        if name in loaded_classes:
+            return (
+                f"{name} or some parts of it are already loaded. "
+                "patch_sklearn() only affects classes imported *after* calling it. "
+                "To retrieve patched entities, make sure to call patch_sklearn() before any import statements from {name}."
```
Comment (Alexsandruss) [inline sklearnex/dispatcher.py:366]: Duplication (see above)

File: sklearnex/tests/test_patching.py [hunk 6]
```diff
@@ -328,3 +328,18 @@ def test_onedal_supported_member(name, member):
     patched = PATCHED_MODELS[name]
     sig = str(inspect.signature(getattr(patched, member)))
     assert "(self, method_name, *data)" == sig
+
+
+def test_check_entity_loaded():
+    # Test when sklearn is not loaded
+    assert check_entity_loaded(modules={}) is None
```
Comment (ethanglaser) [inline sklearnex/tests/test_patching.py:335]: add import for check_entity_loaded

Comment (ahuber21) [reply sklearnex/tests/test_patching.py:335]: What do you mean exactly?

File: sklearnex/dispatcher.py (hunk 5 above)
Comment (ahuber21) [reply sklearnex/dispatcher.py]: Thanks. Simplified it.

File: sklearnex/tests/test_patching.py (hunk 6 above)
Comment (ethanglaser) [reply sklearnex/tests/test_patching.py:335]: CI fails because there is no import of `check_entity_loaded` at top of this file

Comment (ahuber21) [reply sklearnex/tests/test_patching.py:365]: Sometimes the brain stops braining ...

File: sklearnex/dispatcher.py [hunk 7]
```diff
@@ -313,7 +317,61 @@ def get_patch_names():
...
+                        if isinstance(cls, type)
+                    ]
+                )
+        return loaded_classes
+
+    # list of all loaded modules, uses sys.modules per default
+    modules = modules if modules is not None else sys.modules
+
+    # is `name` or anything from sklearn already loaded?
+    loaded = name is None and "sklearn" in modules.keys()
```
Comment (ethanglaser) [inline sklearnex/dispatcher.py:361]: From checking locally it seems that `from sklearnex import patch_sklearn` will yield a true result from `"sklearn" in modules.keys()` (even if nothing is directly imported from sklearn) - therefore the warning is raised in all cases. For example it is raised in our internal conformance checks. Not sure on best alternative here.

**Comment by ahuber21** [comment]
Loading something from `sklearnex` brings in so much stuff from `sklearn` it's really tough to tell if we already have some instances of `sklearn` classes. In my latest commit I tried querying all objects from the garbage collector, but event that's unreliable. On top, it's very slow. I don't see a good way of implementing this feature right now. Closing.

**Pull request closed without merging**